docs/dev/architecture/workflow.md

```diff
@@ -0,0 +1,31 @@
+# Transformation workflow
+
+This document describes the current transformation workflow.
+Older versions are described afterwards.
+
+## Current Flow
+![Flow of transformation, current solution](transformation-flow-flex.png)
+
+##### Which problems have been solved with this design?
+- This design keeps the simplicity (regarding the implementation of the transformation module) of the first design idea and combines it with the flexibility (regarding output files) of the second one.  
+- The plugins will always produce three kinds of output when transforming: First, the target model artifacts itself (e.g.: Cloud Formation Template and Deployment Artifacts).  
+Then, the `instance.properties` text file. It contains keys of required properties at deploy time (note: the values are unset).
+Last, they produce a deployment script `deploy.sh` which containts deployment and orchestration logic. When executing this script, it first sources the `instance.properties` text file, then deploys and orchestrates the app.
+    - This approach enables the user to transform any supported csar, then download the resulting artifacts and deploy at any given time with lightweight dependencies (transformator not required, only dependencies are shell & basic unix tools & targetPlatform-specific tools).
+    - However, the user still has the possibility to declare his application-specific properties and deploy the app via the GUI / CLI. Note that internally, the above mentioned `instance.properties` file is filled with values and the `deploy.sh` script is executed.
```

Comment:
Shell scripts might not be the best solution to provide platform independence. 
Maybe we could use a scripting language like python to execute the scripts. Or the plugin will supply something like a "deployment tool" that inplements the task in java.

Reply:
What bothers me is that if we use scripts (e.g. bash or python) we have to handle missing dependencies. simplest solution would be to tell user what is missing (e.g.: "missing aws-cli in classpath, aborting") and let the user install dependencies on their own. 

If we had a deployment tool written in java for every plugin (client coined this an "executor") I imagine building the plugins could become easier.
However, most important goal then would be to keep this executor lightweight.

Reply:
I think we can expect from a user using our tool to install missing dependencies.

> If we had a deployment tool written in java for every plugin (client coined this an "executor") I imagine building the plugins could become easier.
However, most important goal then would be to keep this executor lightweight.

I thought this was our plan?

Reply:
>I think we can expect from a user using our tool to install missing dependencies.

you're using manjaro and you wouldn't want any script to somehow install further dependencies -- you want that to happen through pacman. on ubuntu its yaourt, on windows there's none -- so how would you handle this..?

> I thought this was our plan?

This was the clients vision after talking about the issue for 10 minutes. I tried to refine the idea even further. Please read the document again. My vision is to let the transform plugins create a `deploy.sh`, and have a single executor in the app which only has to run this script.

Reply:
I meant with a proper doc about what dependencies our tool needs everyone can install the by themselves.

>This was the clients vision after talking about the issue for 10 minutes. I tried to refine the idea even further. Please read the document again. My vision is to let the transform plugins create a deploy.sh, and have a single executor in the app which only has to run this script.

The client also told us if there is enough time at the end that it might be possible that we implement the feature to save the instance of a deployed application. This might not happen, but how do we do that with you `deploy.sh`. Is this logic then implemented in the executor?
I think if we only produce a `deploy.sh` it is not that useful for other use cases.

Reply:
>The client also told us if there is enough time at the end that it might be possible that we implement the feature to save the instance of a deployed application.

I think this is a misunderstanding. With our current flow, that feature is obsolete. I think the client mentioned this when we discussed version 1 of the flow.

Reply:
> I think this is a misunderstanding. With our current flow, that feature is obsolete. I think the client mentioned this when we discussed version 1 of the flow.

This might be true. One thing that is not clear with your idea, do we get only a deploy script and the properties file or also a model? So the transformer produces a `deploy.sh`, `instance.properties`, `transformedmodel.whatever`?

Another thing. If the user needs to perform multiple actions to run our `deploy.sh` why do we not give him a executor that reduces the dependencies? So he wont have to install python or something that understands bash?

Reply:
>One thing that is not clear with your idea, do we get only a deploy script and the properties file or also a model?

See attached note to node 'transform'. A deploy.sh, instance.properties and an undefined amount of target model artifacts are generated by the transform plugin.

> If the user needs to perform multiple actions to run our deploy.sh why do we not give him a executor that reduces the dependencies?

An executor which reduces dependencies implies it contains these dependencies. This is not considered lightweight.

Reply:
> An executor which reduces dependencies implies it contains these dependencies. This is not considered lightweight.

Not using Bash or Python is a dependency less or am I wrong?

Reply:
I'm not sure what kind of executor you mean; but i guess you mean a java executor, which introduces java dependency (i can live with that). This java executor would want to use some target platform specific sdks, e.g. aws sdk for java; to be portable that executor would contain these libs. Such an executor would be quite big (not my understanding of lightweight)

Reply:
i think if the transformation result is not executable with generic tools, the whole idea of having a complete transformation result even without deployment can be given up.

Reply:
> I'm not sure what kind of executor you mean; but i guess you mean a java executor, which introduces java dependency (i can live with that). This java executor would want to use some target platform specific sdks, e.g. aws sdk for java; to be portable that executor would contain these libs. Such an executor would be quite big (not my understanding of lightweight)

Bash or Python might also need some sdks or tools pre installed. And how big is the transformer if his job is to create a executable script?

Reply:
Everyone can read a bash script and modify it if necessary.
Think about the  equivalent of `deploy.sh`. In which format would it be inm if the executor was a java executor? I guess it would be a custom format.

Reply:
If we have a java executor you don't have a custom format you just feed him the transformed model and `instance.properties` file

Reply:
Unfortunately, you need means to communicate the non-transformable aspects of the CSAR to the executor -- he wants to realize these aspects via orchestration.

Reply:
> Unfortunately, you need means to communicate the non-transformable aspects of the CSAR to the executor -- he wants to realize these aspects via orchestration

You are right.

But I don't think this discussion leads to anything so I am out of this.